Avoid unique constraint violations

mikebonnet opened the pull request:
Change inserts for Message, User, and Package objects into INSERT ... SELECTs, which avoids race conditions between existence checking and insert, which can lead to unique constraint violations.

Also fix up the tests so it is possible to test against faitout again.

mikebonnet committed:
make testing with faitout work again

mikebonnet committed:
models.add: use INSERT ... SELECT to avoid errors from duplicate messages

Messages can sometimes be delivered more than once by the broker. When this happens,
the unique constraint on msg_id is violated, and we get errors. This change converts
the plain INSERT into a INSERT ... SELECT that will not insert anything if a message
with a matching msg_id already exists in the database. This will allow datanommer to
handle duplicate messages gracefully.

mikebonnet committed:
use INSERT ... SELECT when creating Users and Packages

A race condition exists in the get_or_create() method between checking for
the existence of a User or Package with a given name, and creating the entry
with that name. In the event that multiple instance of datanommer are
populating the same database with messages, this can lead to unique
constraint violations.

Convert get_or_create() to use INSERT ... SELECT to avoid this race condition.

mikebonnet commented:
/cc @ralphbean

mikebonnet committed:
fix the consumer tests to be aware of the new duplicate message behavior

ralphbean reviewed (approved):
:+1: from me.  :)

@jeremycline, do you want to have a look too?

jeremycline commented on datanommer.models/datanommer/models/__init__.py:
```diff
@@ -128,7 +130,16 @@ def add(envelope):
+                         where(Message.msg_id == msg_id))
+    insert = Message.__table__.insert().from_select(cols, select(vals).\
+                                                    where(not_exists))
+    session.execute(insert)
```
Am I right in assuming that the reason for this approach vs handling an exception is so that everything else in the transaction is committed, even if the message already exists? Is it possible for the message to exist and not the other objects being created here?

mikebonnet replied:
You're correct, it's to avoid a database-level error that would require the transaction to be rolled-back. In this case, it's not possible for a message to exist in the database without the associated user/package. However, if there are two message consumers populating the same database, then user/package creation will race between those two consumers. The INSERT ... SELECT construct does the existence check and insert atomically, avoiding that race condition.

jeremycline replied:
Okay, cool. Why not just handle the exception and roll back the transaction, then? While this approach probably provides a little bit more performance, I'm not certain it's worth the loss of readability.

mikebonnet replied:
This seemed more elegant, rather than triggering constraint violations in the database. Could it be reformatted to improve readability? Would you prefer if it was raw sql rather than the sqlalchemy API?

jeremycline replied:
I would definitely not prefer raw SQL 😄.

In my opinion, it's fine to trigger a constraint violation in the database to determine that the message already exists. After all, that's one of the great features that a database offers. Exceptions are also a great way for this to be communicated and SQLAlchemy's ORM makes good use of exceptions. Bypassing the ORM layer is, to me, only acceptable in special cases where there are huge performance benefits that outweigh the loss of readability and the "surprise" factor.

> However, if there are two message consumers populating the same database, then user/package creation will race between those two consumers. 

This seems odd to me. If there are two workers, why are they both handling the same message? Is it just an exceptional case where a message might be re-transmitted? If so, it seems like performance shouldn't matter much. If not, what's the point of having two workers?

mikebonnet replied:
There are two cases here. One is duplicate message delivery. This should be a rare/exceptional case, but appears to happen in the case of failover/reconnection. Agreed that this *could* be handled by a try/catch, I just chose to do it this way instead. If it's possible to avoid triggering an error condition, it seems like it's sensible to do so.

The other case is two workers processing two different messages, but which both relate to the same user or package. The get_or_create() method had a race condition between checking for the existence of that user/package and inserting it. When there was a collision, the second worker would hit a unique constraint violation, and the message would get lost. Handling this case requires either nested transactions, or the INSERT ... SELECT construct used in Singleton.get_or_create().

jeremycline replied:
> If it's possible to avoid triggering an error condition, it seems like it's sensible to do so.

I don't agree with this approach as I feel it breaks with a very common Python pattern. Error conditions exist to be triggered. It's far better to handle them explicitly than to introduce unnecessary complexity to dodge them.

> The other case is two workers processing two different messages, but which both relate to the same user or package. The get_or_create() method had a race condition between checking for the existence of that user/package and inserting it. When there was a collision, the second worker would hit a unique constraint violation, and the message would get lost.

To me, this is a signal that the transaction doesn't make sense. Nested transactions or a series of transactions seems appropriate.

mikebonnet replied:
I can understand the objection to the duplicate message handling, since that is an exceptional condition. I could rewrite that to do a try/catch. However, I do feel like the approach in get_or_create() is correct, and is the optimal way to handle concurrency when multiple workers are populating the same tables with singleton data.

mikebonnet replied:
I'll note that datanommer already drops out of the SQLAlchemy ORM and into the sql abstraction layer, in the same way this change does. See user_assoc_table and pack_assoc_table.

mikebonnet committed:
skip processing duplicate messages

If the msg_id already exists in the database, an IntegrityError will be
raised on insert. Catch the error and abort processing, because the
message has already been successfully processed once. This should be a
rare occurrence.

jeremycline commented:
I don't agree with the approach used in this pull request. Stepping around the ORM to avoid having the database do its job and check uniqueness constraints introduces unnecessary complexity and reduces maintainability.

My recommended approach is to let the ORM raise the exception, handle it explicitly, and break the transaction up. It's perfectly fine to have a nested transaction or multiple transactions, and that's far clearer to me. Performance should not be a concern since this is not a hot code path. If it is a hot code path, I recommend adjusting the workers that are both trying to insert the same data into the database.

However, I'm not the maintainer of this code. If @ralphbean is happy with this and willing to maintain it, then it's fine by me.

jeremycline commented on datanommer.models/datanommer/models/__init__.py:
```diff
@@ -128,7 +131,14 @@ def add(envelope):
-    session.add(obj)
+    try:
+        session.add(obj)
+        session.flush()
```
Did you mean ``session.commit`` here?

``flush`` just pushes everything to the database's transaction buffer and doesn't actually commit the transaction. You might get an ``IntegrityError`` if the message was already committed to the database before the transaction began, but you definitely won't see one if there's a concurrent transaction that's also adding this message. Whoever commits first wins, but the call to commit later isn't handling the conflict.

mikebonnet replied:
According to http://docs.sqlalchemy.org/en/latest/orm/session_api.html#sqlalchemy.orm.session.Session.flush flush() actually issues the INSERTs to the database, so if a row already exists in the database, it will raise an IntegrityError here, and it's handled appropriately.

jeremycline replied:
It does indeed generate the SQL and issue it to the database, but it does so as part of a transaction:

> Database operations will be issued in the current transactional context and do not affect the state of the transaction, unless an error occurs, in which case the entire transaction is rolled back.

Check out [the committing](http://docs.sqlalchemy.org/en/latest/orm/session_basics.html#committing) documentation and the [transaction lifecycle](http://docs.sqlalchemy.org/en/latest/orm/session_transaction.html#unitofwork-transaction) documentation.

You're going to see ``IntegrityErrors`` when you commit the transaction to the database.

mikebonnet replied:
@jeremycline That is not consistent with the behavior I'm seeing. Remove the "try/except" and run the tests and you'll see an `IntegrityError` raised when `flush()` is called. I believe the doc you quoted there is explaining that `flush()` just issues the statements to the database, but does not issue any `commit` or `rollback`.

pypingou replied:
My experience is similar to the one of @mikebonnet I basically use ``.flush()`` to find out database integrity error (or duplicate error or...) while still giving me the possibility to rollback the transaction later on, which a commit would not allow.

jeremycline replied:
Interesting. This is not the behavior I expected from reading the documentation, but based on the PostgreSQL [transaction isolation](https://www.postgresql.org/docs/current/static/transaction-iso.html) documentation, it makes sense, since it's locking the table for the insert and waiting on concurrent transactions to commit or rollback. My assumption was that was that didn't block and resolved itself at commit-time.

However, I still fail to understand why you're not committing the transaction.

mikebonnet replied:
For the same reason we don't commit after every DML operation, so that everything is consistent before commit. If an error happened later on (when adding the user/package -> message mappings for example), it would throw an error, roll back the entire transaction, and you could potentially retry in the future, when the message is redelivered by the broker (depending on your message ACK settings) for example.

jeremycline replied:
I did not realize you were still relying on the new ``get_or_create`` method. I was expecting your final commit/flush to potentially generate an exception.

Personally, I find this to be more difficult to reason about and I would do it all one way or the other (I favor the ORM), but as I said before I'm not responsible for maintaining this. As long as those who do have that responsibility are satisfied, I'm satisfied.

ralphbean commented:
Alllll right.  Let's do it.
